org.opentosca.bus.management.api.osgievent/src/org/opentosca/bus/management/api/osgievent/MBEventHandler.java

```diff
@@ -157,8 +212,489 @@ public void handleEvent(final Event event) {
+                    && validRelationState.contains(inst.getState())) {
+                    currentlyRunningRelations.add(relation);
+                }
+            }
+        }
+
+        return new ServiceTemplateInstanceConfiguration(currentlyRunningNodes, currentlyRunningRelations);
+    }
+
+    private ServiceTemplateInstanceConfiguration getValidServiceTemplateInstanceConfiguration(AbstractTopologyTemplate topology,
+                                                                                              Map<String, Collection<Long>> nodeIds2situationIds) {
+
+
+        Collection<AbstractNodeTemplate> validNodes = new ArrayList<AbstractNodeTemplate>();
+        Collection<AbstractRelationshipTemplate> validRelations = new ArrayList<AbstractRelationshipTemplate>();
+
+        for (AbstractNodeTemplate nodeTemplate : topology.getNodeTemplates()) {
+            Collection<AbstractPolicy> policies = this.getPolicies(Types.situationPolicyType, nodeTemplate);
+            if (policies.isEmpty()) {
+                validNodes.add(nodeTemplate);
+            } else if (this.isValidUnderSituations(nodeTemplate, nodeIds2situationIds)) {
+                validNodes.add(nodeTemplate);
+            }
+        }
+
+        // check if node set is deployable
+        Collection<AbstractNodeTemplate> deployableAndValidNodeSet =
+            this.getDeployableSubgraph(validNodes, nodeIds2situationIds);
+        for (AbstractRelationshipTemplate relations : topology.getRelationshipTemplates()) {
+            if (deployableAndValidNodeSet.contains(relations.getSource())
+                & deployableAndValidNodeSet.contains(relations.getTarget())) {
+                validRelations.add(relations);
+            }
+        }
+
+        return new ServiceTemplateInstanceConfiguration(deployableAndValidNodeSet, validRelations);
+    }
+
+    private Collection<AbstractNodeTemplate> getDeployableSubgraph(Collection<AbstractNodeTemplate> nodeTemplates,
+                                                                   Map<String, Collection<Long>> nodeIds2situationIds) {
+        Set<AbstractNodeTemplate> validDeploymentSubgraph = new HashSet<AbstractNodeTemplate>(nodeTemplates);
+        Collection<AbstractNodeTemplate> toRemove = new HashSet<AbstractNodeTemplate>();
+
+        for (AbstractNodeTemplate nodeTemplate : nodeTemplates) {
+            Collection<AbstractRelationshipTemplate> hostingRelations = this.getOutgoingHostedOnRelations(nodeTemplate);
+            if (!hostingRelations.isEmpty()) {
+                // if we have hostedOn relations check if it is valid under the situation and is in the set
+                boolean foundValidHost = false;
+                for (AbstractRelationshipTemplate relationshipTemplate : hostingRelations) {
+                    AbstractNodeTemplate hostingNode = relationshipTemplate.getTarget();
+                    if (this.isValidUnderSituations(hostingNode, nodeIds2situationIds)
```

Comment:
Remove `foundValidHost ` and just add the NodeTemplate to the List and break in this if statement?

Reply:
done